src/util/threadpool.h

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2024-present The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or https://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_UTIL_THREADPOOL_H
+#define BITCOIN_UTIL_THREADPOOL_H
+
+#include <sync.h>
+#include <util/string.h>
+#include <util/thread.h>
+#include <util/threadinterrupt.h>
+
+#include <algorithm>
+#include <atomic>
+#include <condition_variable>
+#include <cstddef>
+#include <functional>
+#include <future>
+#include <memory>
+#include <stdexcept>
+#include <utility>
+#include <queue>
+#include <thread>
+#include <vector>
+
+class ThreadPool {
+
+private:
+    Mutex cs_work_queue;
+    std::queue<std::function<void()>> m_work_queue GUARDED_BY(cs_work_queue);
+    std::condition_variable m_condition;
+    CThreadInterrupt m_interrupt;
+    std::vector<std::thread> m_workers;
+
+    void WorkerThread() EXCLUSIVE_LOCKS_REQUIRED(!cs_work_queue)
+    {
+        WAIT_LOCK(cs_work_queue, wait_lock);
+        while (!m_interrupt) {
+            std::function<void()> task;
```

Comment:
Can we modify this to be more generic and return a type? And add logic to collect all returned values into a shared vector which can then be atomically swapped out by an observer? Possibly not in this PR, but if this will be split out into a generic thread pool.